### ISSUE TITLE
Possible to update tenant_id of a DeviceCodeCredential object?

### ISSUE THREAD

**Issue by kuzhao** [Azure/azure-sdk-for-python#31826]
The context is that my program should be able to automatically pick the last AAD tenant in the list for subsequent use, after successful deviceLogin by user. Code snippet:
```python
import azure.identity as identity
credential = identity.DeviceCodeCredential()
    arm_sub_client = arm.SubscriptionClient(credential)
    tenant_id = ''
    for tenant in arm_sub_client.tenants.list():
        tenant_id = tenant.tenant_id

```
Then I tried to update "credential" object's tenant_id:
```python
    if not tenant_id:
        sys.exit('Failed to set the correct tenantId')
    else:
        print(tenant_id)
        credential._tenant_id = tenant_id
```
But it seems arm_sub_client turns a blank eye on this update. The below iterator print code returns an empty sub list:
```python
print([["Subscription_name", "Subscription ID"]] + [
        [sub.display_name, sub.subscription_id]
        for sub in list(arm_sub_client.subscriptions.list())
    ])
```

In contrast, if I hardcode the tenant_id at the very first by
```python
credential = identity.DeviceCodeCredential(tenant_id='xxxx')
```
subscriptions then get listed as intended.

One workaround is to reinitialize "DeviceCodeCredential" object after the first login, and recreate a SubscriptionClient. But subscriptions.list would then trigger another interactLogin which would look silly to the user. 

So, in a nutshell, what is the right way to do a blind deviceCode and then auto fill tenantId?

**Comment by kuzhao** [comment]
Figured out one answer to this problem, along with the potential reason.  
**To make it work**, "arm_sub_client" (or other kinds of AzureClient that make use of credential obj from azure-identity) needs to be re-instantiated.  
**Why?** Merely setting "_tenant_id" field inside credential obj will NOT alter the access token which is already obtained and in use by arm_sub_client. A new client from the modified credential obj will go for a new token under the target tenant.

Fixed code (changed to camelCase for var naming):
```python
    credential = identity.DeviceCodeCredential()
    armSubClient = arm.SubscriptionClient(credential)
    tenantId = ''
    for tenant in armSubClient.tenants.list():
        tenantId = tenant.tenant_id
    armSubClient.close()

    if not tenantId:
        sys.exit('Failed to set the correct tenantId')
    else:
        print(tenantId)
        credential._tenant_id = tenantId
        armSubClient = arm.SubscriptionClient(credential)
    subId = ''
    for sub in armSubClient.subscriptions.list():
        subId = sub.subscription_id
    if not subId:
        sys.exit('Failed to set the subscriptionId')
    else:
        print("SubscriptionId: "+subId)
```
Not sure if there is actually a best practice or more graceful way?

**Comment by xiangyan99** [comment]
Thanks for reaching out.

As you figured out, we cache token to get better performance.

We would recommend using different credential instances for different tenants.
